content/en/blog/_posts/2024-04-15-devsecops-in-action-k8s-secrets.md

```diff
@@ -0,0 +1,562 @@
+```yaml
+  envFrom:
+    - secretRef:
+        name: demo-secret
+```
+
+## Stage-01: Reject environment vars for storing sensitive data
+
+According to OWASP recommendation, it's necessary to use file instead of env.
+That's why it's necessary to make changes to implementation readiness probes
+and deployment manifests.
+
+*[cmd/main.go](https://github.com/efrikin/devsecops-in-action-kubernetes-secrets/blob/96afd59f8a391973a8ed5a9b410915b9e99d4bf4/cmd/main.go#L11-L67)*
+```go
+const (
+	envVar = "DEMO_SECRET__PASSWD_FILE"
+)
+
+...
+
+func fileExist(path string) bool {
+	info, err := os.Stat(path)
+	if err == nil {
+		return !info.IsDir()
+	}
+	if errors.Is(err, os.ErrNotExist) {
+		return false
+	}
+	return false
+}
+
+func httpHandle(w http.ResponseWriter, r *http.Request) {
+
+...
+
+	if envVal, ok = os.LookupEnv(envVar); ok {
+		if fileExist(envVal) {
+			isExist = true
+		}
+	}
+}
+```
+
+Now applications check `DEMO_SECRET__PASSWD_FILE` was defined then application
+checks whether the file exists by value from the environment variable.
+
+*[kustomization/overlays/demo/kustomization.yml](https://github.com/efrikin/devsecops-in-action-kubernetes-secrets/blob/96afd59f8a391973a8ed5a9b410915b9e99d4bf4/kustomization/overlays/demo/kustomization.yml#L25-L45)*
+```yaml
+...
+spec:
+  containers:
+  - envFrom:
+    - secretRef:
+         name: demo-secret
+    volumeMounts:
+    - mountPath: /secrets
+      name: demo-secret
+      readOnly: true
+  volumes:
+  - name: demo-secret
+    secret:
+      optional: true
+      secretName: demo-secret
+...
+```
+
+`spec.containers.0.envFrom` defines all of the Secret's data as container
+environment variables and `spec.volume/spec.containers.0.volumeMounts`
+defines all mountpoints to Pod.
+
+Excellent! Part of recommendations have been completed, but for the full picture
+it's necessary to solve the following issues:
+
+* The secret file is a mountpoint and it's necessary to remove after it usage
+(e.g reading)
+* Kind Secret is insecure that's why it's necessary to minimize its use
+* Additional implementation of environment variable such as:
+`DEMO_SECRET__PASSWD_FILE` can Impede applications update.
+
+## Stage-02: Injection of secret to Pod instead of mountpoint
+
+In order to get rid of Secret as mountpoint it's necessary to use Injection of
+Secret via file to Pod. Often secret manager tools perform these functions.
+Also they provide functions such as: dynamic secrets, database credential
+rotation, automated PKI infrastructure, data encryption and tokenization,
+identity-based access, key management, etc.
+
+In the example below, simulation of getting secret from secure storage and
+delivery to Pod will be done via `spec.initContainers` then
+`spec.containers.0.postStart` hook will clean the file after successful passing
+of readiness probe in order to make file with sensitive data to Pod useless for
+attacker.
```

Comment:
```suggestion
In the following example, we use `spec.initContainers` to simulate fetching a
Secret from an external secret store and delivering it to the Pod. After the Pod
readiness probe succeeds, the `spec.containers.0.postStart` hook cleans the
file so that it's useless to an attacker.
```